File: backend/app/services/analytics_service.py
from app.ml.predictor import ClusterPredictor, get_predictor
from app.schemas.analytics import PredictionRequest, PredictionResponse


class AnalyticsService:
    def __init__(self, predictor: ClusterPredictor | None = None) -> None:
        self.predictor = predictor or get_predictor()

    def predict_cluster(self, payload: PredictionRequest) -> PredictionResponse:
        return self.predictor.predict(payload)

File: frontend/src/app/core/services/analytics.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import {
  PredictionRequest,
  PredictionResponse,
} from '../models/prediction.model';

@Injectable({ providedIn: 'root' })
export class AnalyticsService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiBaseUrl}/analytics`;

  predict(payload: PredictionRequest): Observable<PredictionResponse> {
    return this.http.post<PredictionResponse>(`${this.baseUrl}/predict`, payload);
  }
}
